### PR TITLE
Avoid computing keys multiple times in Cacher

### PR DESCRIPTION
**Pull request by wojtek-t**
This should significantly reduce both cpu-usage and number of allocations in Cacher.

This is a proper follow-up from #30998

@deads2k @liggitt

---

This change is [Reviewable](https://reviewable.kubernetes.io/reviews/kubernetes/kubernetes/35029)

### CODE REVIEW

**Comment by wojtek-t** [comment]
@liggitt - I'm adding you as a reviewer too since you had strong opinions in the previous PR. Feel free to unassign yourself though.

File: pkg/storage/cacher.go [hunk 1]
```diff
@@ -752,7 +752,7 @@ func (c *cacheWatcher) process(initEvents []watchCacheEvent, resourceVersion uin
 	const initProcessThreshold = 500 * time.Millisecond
 	startTime := time.Now()
 	for _, event := range initEvents {
-		c.sendWatchCacheEvent(event)
+		c.sendWatchCacheEvent(&event)
```
Comment (liggitt) [inline pkg/storage/cacher.go]: does this depend on all watchers treating the incoming event as immutable?

File: pkg/storage/watch_cache.go [hunk 2]
```diff
@@ -193,7 +222,7 @@ func (w *watchCache) processEvent(event watch.Event, resourceVersion uint64, upd
 	w.updateCache(resourceVersion, &watchCacheEvent)
 	w.resourceVersion = resourceVersion
 	w.cond.Broadcast()
-	return updateFunc(event.Object)
+	return updateFunc(&storeElement{Key: key, Object: event.Object})
```
Comment (liggitt) [inline pkg/storage/watch_cache.go]: avoid constructing this twice in the function?

File: pkg/storage/watch_cache.go [hunk 3]
```diff
@@ -249,30 +280,56 @@ func (w *watchCache) ListKeys() []string {
 	return w.store.ListKeys()
 }
 
+// Get takes runtime.Object as a parameter. However, it returns
+// pointer to <storeElement>.
```
Comment (liggitt) [inline pkg/storage/watch_cache.go]: why the asymmetry with List()?

File: pkg/storage/watch_cache.go [hunk 4]
```diff
@@ -206,12 +235,14 @@ func (w *watchCache) updateCache(resourceVersion uint64, event *watchCacheEvent)
...
 }
 
+// List returns list of pointers to <storeElement> objects.
 func (w *watchCache) List() []interface{} {
 	w.RLock()
 	defer w.RUnlock()
 	return w.store.List()
 }
 
+// List returns list of pointers to <storeElement> objects.
```
Comment (liggitt) [inline pkg/storage/watch_cache.go]: s/List/WaitUntilFreshAndList/

File: pkg/storage/watch_cache.go [hunk 5]
```diff
@@ -249,30 +280,56 @@ func (w *watchCache) ListKeys() []string {
...
+		return nil, false, fmt.Errorf("couldn't compute key: %v", err)
+	}
+
 	w.RLock()
 	defer w.RUnlock()
-	return w.store.Get(obj)
+	return w.store.Get(&storeElement{Key: key, Object: object})
 }
 
+// Get returns pointer to <storeElement>.
```
Comment (liggitt) [inline pkg/storage/watch_cache.go]: s/Get/GetByKey/

File: pkg/storage/watch_cache.go [hunk 6]
```diff
@@ -72,6 +89,9 @@ type watchCache struct {
 	// Maximum size of history window.
 	capacity int
 
+	// keyFunc is used to get a key in the underyling storage for a given object.
```
Comment (liggitt) [inline pkg/storage/watch_cache.go]: s/underyling/underlying/

File: pkg/storage/cacher.go [hunk 7]
```diff
@@ -524,14 +522,9 @@ func forgetWatcher(c *Cacher, index int, triggerValue string, triggerSupported b
 	}
 }
 
-func filterFunction(key string, keyFunc func(runtime.Object) (string, error), p SelectionPredicate) FilterFunc {
+func filterFunction(key string, p SelectionPredicate) func(string, runtime.Object) bool {
```
Comment (liggitt) [inline pkg/storage/cacher.go]: use `filterObjectFunc` type as return value?

File: pkg/storage/cacher.go [hunk 8]
```diff
@@ -772,7 +772,7 @@ func (c *cacheWatcher) process(initEvents []watchCacheEvent, resourceVersion uin
 		}
 		// only send events newer than resourceVersion
 		if event.ResourceVersion > resourceVersion {
-			c.sendWatchCacheEvent(event)
+			c.sendWatchCacheEvent(&event)
```
Comment (liggitt) [inline pkg/storage/cacher.go]: same question here, what happens if a watcher mutates an event (or object in an event)?

File: pkg/storage/cacher.go (hunk 7 above)
Comment (wojtek-t) [inline pkg/storage/cacher.go]: Sure - done

File: pkg/storage/cacher.go (hunk 1 above)
Comment (wojtek-t) [inline pkg/storage/cacher.go]: Yes (and this is true assumption today). I added explicit comment for that method.

File: pkg/storage/cacher.go (hunk 8 above)
Comment (wojtek-t) [inline pkg/storage/cacher.go]: As above - added comment to "sendWatchCacheEvent" method.

File: pkg/storage/watch_cache.go (hunk 6 above)
Comment (wojtek-t) [inline pkg/storage/watch_cache.go]: done

File: pkg/storage/watch_cache.go (hunk 4 above)
Comment (wojtek-t) [inline pkg/storage/watch_cache.go]: done

File: pkg/storage/watch_cache.go (hunk 5 above)
Comment (wojtek-t) [inline pkg/storage/watch_cache.go]: done

**Comment by wojtek-t** [comment]
@liggitt - all comments applied; PTAL

File: pkg/storage/cacher.go (hunk 8 above)
Comment (liggitt) [inline pkg/storage/cacher.go]: ok, just realized sendWatchCacheEvent copies the object and constructs a watch.Event per watcher, so that's fine

**Comment by liggitt** [comment]
LGTM, any numbers on how this helped performance?

**Comment by wojtek-t** [comment]
>
LGTM, any numbers on how this helped performance?

I did only some small experiments (i.e. with 10 nodes) and it almost eliminated "NamespaceKeyFunc" from the profile (which is usually around 10% of cpu profile and 10% of allocations).
Running larger experiments manually is pretty painful (we are running them continuously from head) so I will give you better numbers few hours after this is merged.

**Comment by wojtek-t** [comment]
@liggitt - so to confirm, even in very large tests the "NamespaceKeyFunc" was completely eliminated from profiles; also the 99% of latency of listing pods is significantly better in our tests (the difference is around 1.5x).

**Pull request merged**